### PR TITLE
network-generator: support setting NamePolicy=

### LINKED ISSUE #16296: support controlling NIC device NamePolicy via kernel arguments
**Is your feature request related to a problem? Please describe.**

No problem.

It would be nice to be able to control my NIC naming via a karg. 

**Describe the solution you'd like**

I can already disable predictable names via `net.ifnames` and I can already control the naming scheme via `net.naming-scheme`. It would be nice to be able to control the `NamePolicy` as defined in `/usr/lib/systemd/network/99-default.link` via a karg as well.

An example here would be useful. I run Fedora CoreOS. Fedora CoreOS starts as a disk image and configuration gets applied via Ignition on first boot. If I would like to change my `NamePolicy` to `mac` such that my interfaces are named like `enx123456789abc` then I have to boot the image, write a file into `/etc/systemd/network/99-default.link` and reboot. It would be more convenient if I could control this from the start with a karg.

Maybe this was considered before and decided against for good reason.

**Describe alternatives you've considered**

The alternative is what we have today :)

### PR DESCRIPTION
**Pull request by yuwata**
Closes #16296.

### CODE REVIEW

File: src/network/generator/network-generator.c [hunk 1]
```diff
@@ -276,28 +281,48 @@ static int link_new(
...
+        }
+
+        if (!filename && !hw_addr_is_null(mac)) {
+                filename = strdup(HW_ADDR_TO_STR_FULL(mac, HW_ADDR_TO_STRING_NO_COLON));
+                if (!filename)
+                        return -ENOMEM;
+        }
+
+        if (!filename) {
+                filename = strdup("default");
```
Comment (keszybz) [inline src/network/generator/network-generator.c]: Maybe combine those two:
```C
if (!filename) {
     filename = strdup(hw_addr_is_null(mac) ? "default" :
 HW_ADDR_TO_STR_FULL(mac, HW_ADDR_TO_STRING_NO_COLON));
                if (!filename)
                        return -ENOMEM;
        }
```

**Review by keszybz — COMMENTED** [review]
Looks good. Doesn't this need an update in the docs somewhere?

I think we can merge this after the release.

**Comment by yuwata** [comment]
@keszybz Thank you for the review. Updated. Now, the new command line argument is briefly documented.

**Comment by keszybz** [comment]
Looks reasonable.

@dustymabe, does this work for you?

**Comment by yuwata** [comment]
Indeed. the new argument name should be changed.

> Is this systemd-networkd specific? For my particular use, we use NetworkManager.

No. It creates .network, .netdev, and .link files. And .link files are for udevd. Hence, the first commit of this PR is important.

**Comment by yuwata** [comment]
Renamed to `net.ifname-policy=`. PTAL.

**Comment by dustymabe** [comment]
I didn't review the code, but the docs changes look reasonable for what I would expect. Thanks!

**Comment by yuwata** [comment]
I noticed that the service is not built when networkd is disabled.

@keszybz WDYT about moving the logic to generate .link files into systemd-udevd? And make network-generator only generates .network and .netdev files. One reason I introduced network-generator as a separated executable was that /run/systemd/network is not writable by networkd as which runs with systemd-network user. But, udevd runs with root privilege, so I guess the feature to generate .link files can be moved to udevd.

**Comment by keszybz** [comment]
I think it's nice that the generator is separate from the daemons. I'm not too familiar with the layout of the code, but what about just making the generator independent of networkd and udev and building it unconditionally? It seems the code is mostly independent of networkd, so it could be built even if networkd itself is disabled. If the link parts were moved into udevd, udevd would either generate files that it would consume itself, or it would need to use the cmdline information directly. But I think it's nice for debugging that the intermediate step of .link files is used.

**Comment by dustymabe** [comment]
> If the link parts were moved into udevd, udevd would either generate files that it would consume itself,

i.e. a `systemd-udevd-generator`? Would that be horrible?

**Comment by yuwata** [comment]
OK. Now network-generator is built unconditionally.

**Comment by keszybz** [comment]
> i.e. a `systemd-udevd-generator`? Would that be horrible?

Yes ;)



> OK. Now network-generator is built unconditionally.

Thanks. I think this DTRT.

The ppc64el failure should be fixed by #21784.
centos ci/(centos stream 8) is the test_macvlan stuff, that was also appearing in other places.
So those failures don't look relevant.

I think we should merge this now. The first commit fixes a potential race and I think regressions are unlikely with the other commits, especially if the new functionality is not used.

**Pull request merged**